core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java

```diff
@@ -1012,4 +1028,49 @@ static void ensureNotNull(Object value, String message) {
             throw new IllegalArgumentException(message);
         }
     }
+
+    static void ensureNoSelfReferences(Object value) {
+        ensureNoSelfReferences(value, Collections.newSetFromMap(new IdentityHashMap<>()));
+    }
+
+    private static Object ensureNoSelfReferences(final Object value, final Set<Object> ancestors) {
+        if (value != null) {
+
+            Iterator it = null;
+            if (value instanceof Map) {
+                it = ((Map) value).values().iterator();
+            } else if ((value instanceof Iterable) && (value instanceof Path == false)) {
+                it = ((Iterable) value).iterator();
+            } else if (value instanceof Object[]) {
+                it = new Iterator<Object>() {
+                    int index = 0;
+                    final Object[] array = (Object[]) value;
+
+                    @Override
+                    public boolean hasNext() {
+                        return index < array.length;
+                    }
+
+                    @Override
+                    public Object next() {
+                        return array[index++];
+                    }
+                };
+            }
```

Comment:
otherwise I would be happy to just return in the else block and remove the instanceof check in the while loop below